### DISCUSSION TITLE
Assistance with “IllegalArgumentException: Points of LinearRing do not form a closed linestring” Error in Momepy Tessellation

### DISCUSSION THREAD

**Discussion by mwahyur46** [pysal/momepy#663] [Q&A]
Dear Martin,
I'm currently working on a research project utilizing Momepy's Tessellation function (`momepy.Tessellation` or `momepy.morphological_tessellation`), which will serve as an additional feature besides morphological characters and land use classification generated from remote sensing data for a machine learning model to map settlement types. Although I have conducted geometric and topological corrections on my polygons to ensure they are ready for tessellation, I still encounter the following error:
`GEOSException: IllegalArgumentException: Points of LinearRing do not form a closed linestring`

Additionally, when attempting tessellation, I am prompted to re-project the geometries to a projected CRS, despite having already set them to UTM Zone 49S.

Here’s a summary of my workflow:

1. Polygon Validation: I used shapely to verify that all geometries are closed and valid.
2. CRS Verification: The data is projected in UTM Zone 49S, which should meet tessellation requirements.
3. Closing Gaps: I applied momepy.close_gaps to further refine the geometries, but this didn’t prevent the error.

If possible, would you be willing to review the building polygon file I processed (sourced from Google Open Building) to help identify any additional issues?

Thank you.
[Building_Google Open Building.zip](https://github.com/user-attachments/files/17610916/Building_Google.Open.Building.zip)
![image](https://github.com/user-attachments/assets/b2d59f7d-6e19-4a3d-b212-2ff37cbb611e)

**Comment by jGaboardi** [comment]
@mwahyur46 

Can you provide the following:
* a [minimum reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example#:~:text=In%20computing%2C%20a%20minimal%20reproducible,to%20be%20demonstrated%20and%20reproduced.)
* versions of (at least) the major spatial packages you are working with (including `momepy` itself)
* Python version and OS

This should all be provided in `copy-paste`-able text. It is nearly impossible to attempt to debug from screenshots.

What have you attempted in terms of trying to identify the offending polygon(s) themselves (assuming the issue is actually with the data). I would perhaps start with dividing the data into progressivley smaller chunks to determine if the operations can be performed successfully with subsets.

**Reply by mwahyur46** [reply]
Thanks for the response James, here I try to provide the code that I used to correct the geometry/topology of the building and create the tessellation:

```
import geopandas as gpd
from shapely.geometry import Polygon
from shapely.ops import unary_union

def close_ring(geom):
    """Ensure polygons have closed rings."""
    if isinstance(geom, Polygon) and not geom.exterior.is_ring:
        return Polygon(list(geom.exterior.coords) + [geom.exterior.coords[0]])
    return geom

def validate_and_close_polygons(gdf):
    """Validate geometry and close polygons if necessary."""
    gdf['geometry'] = gdf['geometry'].apply(lambda geom: geom if geom.is_valid else geom.buffer(0))
    gdf['geometry'] = gdf['geometry'].apply(close_ring)
    return gdf

def remove_overlaps(gdf, threshold=0.5):
    """Remove overlaps between polygons above a certain threshold."""
    overlaps = []
    for i, geom1 in enumerate(gdf.geometry):
        for j, geom2 in enumerate(gdf.geometry):
            if i < j and geom1.intersects(geom2):
                intersection = geom1.intersection(geom2)
                overlap_ratio = intersection.area / min(geom1.area, geom2.area)
                if overlap_ratio >= threshold:
                    overlaps.append(gdf.iloc[[i, j]])

    if overlaps:
        overlap_gdf = gpd.GeoDataFrame(pd.concat(overlaps, ignore_index=True), crs=gdf.crs)
        overlap_union = unary_union(overlap_gdf['geometry'])
        overlap_union_gdf = gpd.GeoDataFrame(geometry=[overlap_union], crs=gdf.crs)
        gdf = gpd.overlay(gdf, overlap_union_gdf, how="difference")
    
    return gdf

# Example usage for MRE
input_path = "Data/Building.geojson"
gdf = gpd.read_file(input_path)
gdf = validate_and_close_polygons(gdf)
gdf = remove_overlaps(gdf)

```

Then, the following is the code for creating tessellation from the corrected building polygons:
```
import geopandas as gpd
import momepy
from shapely.validation import make_valid

def convert_to_utm(gdf, utm_zone=32749):
    """Convert data to a projected UTM CRS if necessary."""
    if gdf.crs.is_geographic:
        gdf = gdf.to_crs(epsg=utm_zone)
    return gdf

def validate_geometries(gdf):
    """Validate and repair geometries."""
    gdf['geometry'] = gdf['geometry'].apply(make_valid)
    return gdf

def create_tessellation(building_path, unique_id='bID'):
    """Generate tessellation from building footprints."""
    gdf = gpd.read_file(building_path)
    gdf = convert_to_utm(gdf, utm_zone=32749)
    gdf = validate_geometries(gdf)
    tessellation = momepy.Tessellation(gdf, unique_id, shrink=0.4, segment=0.5)
    return tessellation.tessellation

# Example usage
building_path = "Data/Building.geojson"
tessellation = create_tessellation(building_path)

```

This is my version of Python and OS. and just in case, you need a sample of the building polygon I used from Google Open Building and OSM, I tried both but the problem remains the same. So, my assumption is that either my code, or the Python version is causing this.

- Python version: 3.11.7
- Operating System: Windows 11
- GeoPandas version: 0.14.2
- Momepy version: 0.7.2
- Shapely version: 2.0.6
- Pandas version: 2.2.2
- NumPy version: 1.26.4

[Building_Google Open Building and OSM.zip](https://github.com/user-attachments/files/17614838/Building_Google.Open.Building.and.OSM.zip)
